[READY] Json formatter.

kevingranade opened the pull request:
Leverages the DDA internal json library to format json instead of an external perl script.
This avoids the ordering problem the perl script works around by manually specifying orderng and wraping of elements individually.

Remaining issues
- [x] Fix floating point precision, the current setting isn't suitable for manually-edited configuration.
- [x] Add whitelist support so it can be adopted piecemeal.
- [x] Add CGI support to take over for the web linter.
- [x] Add to travis.yml and possibly Gorgon to perform regression testing.

Coolthulhu commented on data/json/harvest.json:
```diff
@@ -2,16 +2,11 @@
-      { "drop": "stomach" },
-      { "drop": "bone", "base_num": [ 1, 2 ], "scale_num": [ 0.4, 0.5 ], "max": 4 }
-    ]
+    "entries": [ { "drop": "fish", "scale_num": [ 0.200000, 0.400000 ], "max": 2 }, { "drop": "offal", "base_num": [ 0, 1 ], "scale_num": [ 0.200000, 0.400000 ], "max": 1 }, { "drop": "stomach" }, { "drop": "bone", "base_num": [ 1, 2 ], "scale_num": [ 0.400000, 0.500000 ], "max": 4 } ]
```
That forced unwrap is terrible

kevingranade replied:
yea it's a problem, any suggestions for how to address it?

Coolthulhu replied:
The perl script allows configuring whether to break or not. Replicating this would be the most configurable option.
If loading options is too much work, it may be easier to first output the result to string and measure the line length. In case of line length exceeding some constant, the output for a given object would be regenerated, this time with breaks.

format.conf will almost certainly result in less work in the end. Hardcoded formatting - even relatively smart one - may run into hard to predict edge cases where it will generate giant blocks of indents with syntax or 200+ character lines.
For example json mapgen entries, `iuse_actor`s, inline monster attacks - those are all complex and big. It's not feasible to cover those cases with a simple, predictable formatter without having someone specify the rules explicitly.
Predictability is important because unpredictable formatting results in those extra "format x.json" commits.

Coolthulhu commented on data/mods/More_Survival_Tools/furniture.json:
```diff
@@ -7,13 +7,7 @@
-      "sound_fail": "whump!",
-      "items": [ { "item": "rock", "count": [ 5, 15 ] } ]
-    },
+    "bash": { "str_min": 15, "str_max": 105, "sound": "crash!", "sound_fail": "whump!", "items": [ { "item": "rock", "count": [ 5, 15 ] } ] },
```
Same here

ZhilkinSerg commented:
So, format.conf wont' be used anymore?

BevapDin commented on Makefile:
```diff
@@ -953,6 +956,15 @@ else
 endif
 
+style-json: new_json_whitelist json_formatter
+	cat new_json_whitelist | xargs -L 1 ./json_formatter
```
Useless use of `cat`. Can't you use the shells build-in mechanism (`xargs <file`) or `xargs` option `-a` for this?

kevingranade replied:
Sure thing, -a looks best.

BevapDin commented on Makefile:
```diff
@@ -953,6 +956,15 @@ else
+	cat new_json_whitelist | xargs -L 1 ./json_formatter
+
+style-all-json: json_formatter
+	find data -name "*.json" | xargs -L 1 ./json_formatter
```
And this should use `-print0` in the `find` command and `-0` as option to `xargs` to handle paths with white spaces correctly.

kevingranade replied:
Meh, I almost want to make paths with whitespaces break because why the hell would you do that, but sure.

BevapDin commented on src/json.cpp:
```diff
@@ -1604,27 +1610,49 @@ void JsonOut::write_member_separator()
+        indent_level += 1;
+        // Wrap after top level object and array opening.
+        if( indent_level <= 2 ) {
+            stream->put('\n');
```
This doesn't seem to be astyled: there are spaces missing around the `'\n'`.

kevingranade replied:
json.cpp isn't in general, but yea I'll make sure all my additions are.

BevapDin commented on tools/json_tools/format/format.cpp:
```diff
@@ -0,0 +1,88 @@
+    if( filename.empty() ) {
+        std::cout << out.str() << std::endl;
+    } else {
+        std::ofstream fout( filename, std::ios::binary | std::ios::trunc );
```
This should only be done when the strings differ (in the else branch of the below if). That will keep the input file unchanged in the file system, making it simple for diff tools / git.

kevingranade replied:
Great idea.

BevapDin commented on tools/json_tools/format/format.cpp:
```diff
@@ -0,0 +1,88 @@
+        bool tf = jsin.get_bool();
+        jsout.write( tf );
+    } else {
+        // Error?
```
Shouldn't this branch complain loudly, at least write a message to `stderr`?

kevingranade replied:
It's a little tricky to write something meaningful, but I think I managed something.

kevingranade commented:
> So, format.conf won't be used anymore?

Yes, this approach doesn't require annotating all the entries by hand.

The floating point formatting was surprisingly difficult to get to work right. There's a bizarre and gaping hole in c++'s support for what I would presume is the most common human-readable format for writing floating point numbers (digits with no leading 0s, mandatory decimal followed by at least one digit, no trailing 0s otherwise).  So I had to hack together some raw string manipulation to make it work as expected.

kevingranade commented:
Just hapened to run across this: http://journal.stuffwithstuff.com/2015/09/08/the-hardest-program-ive-ever-written/
An extrordinarally full-featured code formatter implementation for the dart programming language.  This kind of approach is almost certainly not necessary here, but worth noting.

Back on target, the change since the last version is I tried a simple tweak to wrapping heuristics.
Basically, object elements are line wrapped, but arrays aren't.
```
{
  "id": "type",
  "nested-array": [ 1, 2, 3, 4, 5 ],
  "nested-object": {
    "int": 1,
    "doubly-nested-array": [ 1, 2, 3, 4, 5 ],
    "doubly-nested-object": {
      "elem": "foo"
    },
    "object-array": [ {
      "type": "foo",
      "val": 1,
      "nested-array": [ 1, 2, 3, 4, 5 ],
      "nested-object": {
        "foo": "bar"
        } ]
}
```
Note that penultimate line, that's representative of an actual bug, but I think I can sort it out.
The tradeoff here is we abandon a lot of potential inlining, but I think this is a ruleset that can be line-length agnostic, so we don't need any data structure or backtracking support in the formatter.
Let me know if there are significant regressions you're aware of, but I think this is a pretty strong contender.

kevingranade commented on data/json/recipes/ammo/shot.json:
```diff
@@ -43,10 +43,7 @@
-      [ [ "gunpowder", 6 ], [ "chem_black_powder", 6 ] ],
-      [ [ "combatnail", 10 ] ]
-    ]
+    "components": [ [ [ "gunpowder", 6 ], [ "chem_black_powder", 6 ] ], [ [ "combatnail", 10 ] ] ]
```
This change is questionable, might need to do something about multiply-nested arrays.

Coolthulhu replied:
Then entries in main array should wrap if:
* Nested arrays wrap
* There are size > 1 nested arrays

The main problem here is the number of `[]`. When all nested arrays are size 1, this is not a problem because the meaning is still obvious from context.

kevingranade replied:
As with https://github.com/CleverRaven/Cataclysm-DDA/pull/21612/files#diff-b845aefebd7df667ac2f080c259bfc5eR23
the innermost element should be an object, yielding:
```
 "components": [ {
    "gunpowder": 6,
    "chem_black_powder": 6
  }, {
    "combatnail": 10
  }
]
```

kevingranade commented on data/mods/blazemod/blaze_other.json:
```diff
@@ -21,7 +21,7 @@
     "price": 6000,
     "displacement": 60,
-    "faults": []
+    "faults": [  ]
```
Might want to preserve empty arrays not containing whitespace.

kevingranade commented on data/json/items/ammo/40.json:
```diff
@@ -20,10 +20,7 @@
-      "damage": -3,
-      "pierce": 6
-    }
+    "relative": { "damage": -3, "pierce": 6 }
```
This isn't styled correctly, it's a commit management issue, not the formatter.

Coolthulhu commented on data/json/harvest.json:
```diff
@@ -2,16 +2,27 @@
+      "drop": "fish",
+      "scale_num": [ 0.2, 0.4 ],
+      "max": 2
+    }, {
```
This needs an indent. It looks like it belongs to the main object, not `entries`.

kevingranade replied:
Agreed, that's a bug.

kevingranade replied:
Hmm, this is a lot trickier to address than expected.  I think the best we can do is:
```
"entries": [ {
    "drop": "fish",
    "scale_num": [ 0.2, 0.4 ],
    "max": 2
  }
]
```

Coolthulhu commented on data/json/harvest.json:
```diff
@@ -2,16 +2,27 @@
+      "drop": "bone",
+      "base_num": [ 1, 2 ],
+      "scale_num": [ 0.4, 0.5 ],
+      "max": 4
```
Arrays containing only small objects should have each object wrap

Coolthulhu commented on data/json/items/ammo/300.json:
```diff
@@ -32,7 +32,11 @@
-    "extend": { "effects": [ "RECYCLED" ] },
-    "delete": { "effects": [ "NEVER_MISFIRES" ] }
+    "extend": {
+      "effects": [ "RECYCLED" ]
```
This should be able to stay on one line

kevingranade replied:
There's no way to do that without backtracking.

Coolthulhu replied:
Requiring the formatter to be one-pass is a huge restriction that may make it unusable.

kevingranade replied:
Aggressive inlining is a nice-to-have at best and an antipattern at worst.

As that is currently the sole reason to make the formatter significantly more complex, I'm not inclined to invest the time to implement it.  If you want to write a backtracking formatter that does smart line wrapping and doesn't require external annotation, feel free.

Coolthulhu commented on data/json/items/gun/44.json:
```diff
@@ -20,11 +20,7 @@
-      [ "45", [ "m1911mag", "m1911bigmag" ] ],
-      [ "9mm", [ "glockmag", "glockbigmag" ] ]
-    ]
+    "magazines": [ [ "44", [ "deaglemag" ] ], [ "45", [ "m1911mag", "m1911bigmag" ] ], [ "9mm", [ "glockmag", "glockbigmag" ] ] ]
```
This should wrap at each ammo type. Forcing it all on one line hurts readability due to `[]` spam.

kevingranade replied:
Hmm, I think the real problem here is that this is an array of arrays of arrays in the first place, it would wrap nicely if it were:
```
"magazines" : {
  "44" : ["deaglemag"],
  "45": [ "m1911mag", "m1911bigmag" ]
}
```

Coolthulhu replied:
That would require it not to be a proper, well defined object though.
It would be very non-standard to have it use data as member names.

kevingranade replied:
A json object is just a key/value map, we just happen to have a lot of objects that map directly to game objects, but it's in no way nonstandard to have a data/data map.

Coolthulhu commented on data/json/requirements/toolsets.json:
```diff
@@ -18,13 +15,19 @@
+    "qualities": [ {
+      "id": "SEW",
+      "level": 1
+    } ]
```
Each of those objects really needs to stay on one line.

BevapDin commented on tools/json_tools/format/format.cpp:
```diff
@@ -0,0 +1,108 @@
+        bool tf = jsin.get_bool();
+        jsout.write( tf );
+    } else {
+        std::cerr << "Encountered unrecognized json element \"";
```
The JSON parser also understand `null` elements - see `JsonIn::test_null`.

kevingranade replied:
Added, missed it since we never use it.

kevingranade commented:
That's not an error, that's 15 THOUSAND lines removed, the examples directory had an absolutely stupid amount of json packed in it as a "regression test".

ZhilkinSerg commented:
Is there currently any sense in adding more linting rules to perl formatter?

kevingranade commented:
> Is there currently any sense in adding more linting rules to perl formatter?

No, unless someone spots a significant issue with this PR, the Perl formatter is going away very soon.

ZhilkinSerg commented:
> No, unless someone spots a significant issue with this PR, the Perl formatter is going away very soon.

So, when it lands it will be possible to lint files locally using cdda binary and/or online via cgi?

Coolthulhu commented on data/json/recipes/chem/other.json:
```diff
@@ -26,11 +27,12 @@
-      [ [ "water", 1 ], [ "water_clean", 1 ] ],
-      [ [ "bleach", 1 ], [ "lye_powder", 100 ], [ "bone", 20 ], [ "bone_human", 20 ], [ "bone_tainted", 40 ], [ "cured_hide", 2 ], [ "pine_bough", 10 ], [ "raw_dandelion", 20 ], [ "biollante_bud", 1 ], [ "ant_egg", 1 ], [ "slime_scrap", 1 ] ]
-    ]
+    "components": [ [ [ "water", 1 ], [ "water_clean", 1 ] ], [ [ "bleach", 1 ], [ "lye_powder", 100 ], [ "bone", 20 ], [ "bone_human", 20 ], [ "bone_tainted", 40 ], [ "cured_hide", 2 ], [ "pine_bough", 10 ], [ "raw_dandelion", 20 ], [ "biollante_bud", 1 ], [ "ant_egg", 1 ], [ "slime_scrap", 1 ] ] ]
```
This line is well above tolerable length. Made worse by tons of braces.

kevingranade replied:
I already addressed this https://github.com/CleverRaven/Cataclysm-DDA/pull/21612#pullrequestreview-57416540

Coolthulhu replied:
"Maybe someday" would work if the PR didn't also get rid of the only formatter capable of outputing sane results for such a recipe.
If the formatter is only capable of formatting very simple entries, it is not ready to replace the old, complete formatter. Instead, it should have its separate whitelist, with the more extensive one still being covered by the formatter which can handle them.

Coolthulhu commented:
> No, unless someone spots a significant issue with this PR, the Perl formatter is going away very soon.

I have a huge one I mentioned before: mapgen.
Your examples don't include mapgen, and judging from existing examples, the new formatter is most likely intolerably inadequate at it.

Also complex recipes. For example, `"sweetbread"`. Those are likely to get totally unreadable to the point where the formatter will only work for the simplest cases. Even the simple cases look bad already.

kevingranade commented:
Poked at limited backtracking and it wasn't quite as much of a pain as anticipated.
I'm fairly sure this adresses both the mapgen and recipe issues, as well as most of the "should be inlined" instances.

Coolthulhu commented on data/json/items/classes/gun.json:
```diff
@@ -33,25 +33,61 @@
+      [ "sights", 1 ],
+      [ "stock", 1 ],
+      [ "underbarrel", 1 ]
+    ]
```
This does look bad. If it was just the base abstract guns it would be tolerable, but there are tons of guns not using this system.

kevingranade replied:
There is no way that wrapping the line off the edge of the screen (225 characters!) is more readable than writing one element per line.

Coolthulhu commented on data/json/items/vehicle/engine.json:
```diff
@@ -13,14 +13,31 @@
+      "fault_engine_pump_diesel",
+      "fault_engine_pump_water",
+      "fault_engine_starter"
+    ]
```
This looks bad, but I guess it can be tolerated, as there aren't many engines in mainline.

Coolthulhu commented on data/json/mapgen/antique_store.json:
```diff
@@ -76,13 +76,48 @@
+              "x": [ 4, 9 ],
+              "y": [ 10, 10 ],
+              "chance": 70
+            },
```
This looks horrible. Mapgens are some of the most object-heavy objects and changing every item entry to at least 5 lines is going to hurt readability a lot.

kevingranade replied:
This isn't remotely readable either way. If you can come up with a way to improve it feel free, but trying to forcibly inline objects to conserve lines (what I have to assume you want to happen since you didn't offer a suggestion) isn't a significant improvement.

Coolthulhu replied:
Old version is clearly readable:
* One entry - one line
* "What to place" is in the same place
* No padding with needless newlines
* Everything can be seen at once. Even a big list will not require scrolling.
* Will not bloat diff size to incredible numbers

Coolthulhu commented on data/json/vehicles/vehicles.json:
```diff
@@ -5,14 +5,19 @@
+        "x": 0,
+        "y": 0,
+        "part": "frame_vertical_2"
+      }
```
Another huge red flag. This file only has the dummy vehicles.
`data/json/vehicles/cars.json` isn't very readable at the moment, but after this formatting, it would look worse than an unformatted freeform mess.

kevingranade commented:
If you have any constructive feedback post it now, otherwise I'm merging this tonight.

Coolthulhu commented:
* No real (not 0 element dummy) vehicle example to show how hard it breaks with vehicles. SPOILERS: It breaks very, very hard with vehicles
* Monster groups, like data/json/monstergroups.json - forced wrapping at each member will make this moderately ugly file unreadable. And not just this file - any big monster group. This one just happen to have the most of big monster groups.
* Item groups, like data/json/item_groups.json - mix of arrays and objects
* Item spawns in mapgen. They're clearly worse off. Their variable newline count makes the whole thing significantly worse.

If you force this incomplete thing in as a replacement for the actually working formatter, the two first types will either only be formatted by hand (why have a formatter then?) or be written as write-once-read-never messes that no one ever bothers to fix beyond number tweak (if that) because it's about as human-readable as minified javascript.
The idea of ever having the whole data/json directory formatted and human-readable at the same time goes out of the window.

Unless someone writes an external tool to manage specific json types. Hilarious idea: being forced to use a tool to avoid the torture of having to deal with a different tool, which in turn has replaced the third tool, which existed to (successfully) make work easier and the results more readable.
